PerformanceMonitor.bsv

```diff
@@ -0,0 +1,163 @@
+    vec_rg_counter <- replicateM(mkCRegU(2));
+  // Which event the corresponding counter will count
+  Vector #(ctrs, Reg #(Bit #(TLog #(evts))))
+    vec_rg_event_sel <- replicateM(mkRegU);
+
+  Reg #(Bit #(ctrs)) rg_ctr_inhibit <- mkReg(0);
+  Wire #(Bit #(ctrs)) wr_overflow <- mkDWire(0);
+
+  // Since [i] is not a function, cannot map it to a Vector
+  function a_type sel_snd_port (Array# (a_type) arr) = arr[1];
+
+  // Should be called every cycle with vector of how many times each event
+  // occured in the given cycle. The maximum is 2^ctrW, but an event 'should'
+  // never happen so many times in a cycle:
+  // So map zeroExtend to the vector before passing it to this method
+  method Action send_performance_events (Vector #(evts, Bit #(ctrW)) events);
+    Bit #(ctrs) overflow = 0;
+    for (Integer i = 0; i < ctrs; i = i + 1) begin
+      // Get the number of times the selected event fired this cycle
+      Bit #(ctrW) event_count = 0;
+      if (vec_rg_event_sel[i] <= fromInteger(evts - 1))
+      event_count = events[vec_rg_event_sel[i]];
+
+      // Check that the given counter is not inhibited
+      Bool inhibit = unpack(rg_ctr_inhibit[i]);
+      // Count event
+      if (!inhibit) begin
+        Bit #(TAdd #(ctrW, 1)) count_sum = zeroExtend (vec_rg_counter[i][0])
+                                          + zeroExtend (event_count);
+        vec_rg_counter[i][0] <= truncate (count_sum);
+        overflow[i] = truncateLSB (count_sum);
+      end
+
+      // $display("<time %0t, StatCounters> Counter #%2d Counting event %0d: Adding %0d to %0d (Inhibit %0d)",
+      //   $time, i, vec_rg_event_sel[i], event_count, vec_rg_counter[i][0], inhibit);
+    end
+    wr_overflow <= overflow;
+    // $display("-------------------------CLK-------------------------");
+  endmethod
+
+  method Vector #(ctrs, ReadOnly #(Bit #(ctrW))) read_counters =
+    map (compose (regToReadOnly, sel_snd_port), vec_rg_counter);
+
+  method Vector #(ctrs, ReadOnly #(Bit #(TLog #(evts)))) read_ctr_sels =
+    map (regToReadOnly, vec_rg_event_sel);
+
+  method ReadOnly #(Bit #(ctrs)) read_ctr_inhibit =
+    regToReadOnly (rg_ctr_inhibit);
+
+  method ReadOnly #(Bit #(ctrs)) read_ctr_overflow =
+    regToReadOnly (wr_overflow);
+
+
+  method write_counter (Bit #(TLog #(ctrs)) idx) =
+    vec_rg_counter[idx][1]._write;
+
+  method write_ctr_sel (Bit #(TLog #(ctrs)) idx) =
+    vec_rg_event_sel[idx]._write;
+
+  method write_ctr_inhibit = rg_ctr_inhibit._write;
+
+endmodule
+
+// For a synthesize module, the values must be fixed
+`ifndef NO_OF_CTRS
+`define NO_OF_CTRS 29
+`endif
+`ifndef COUNTER_WIDTH
+`define COUNTER_WIDTH 64
+`endif
+`ifndef NO_OF_EVTS
+`define NO_OF_EVTS 96
+`endif
+
+(* synthesize *)
+module mkPerfCounters (PerfCounters_IFC #(`NO_OF_CTRS, `COUNTER_WIDTH, `NO_OF_EVTS));
+  PerfCounters_IFC #(`NO_OF_CTRS, `COUNTER_WIDTH, `NO_OF_EVTS) perf_counters <- mkPerfCounters_Core;
+  return perf_counters;
+endmodule
```

Comment:
This arbitrary instantiation doesn't belong in BlueStuff; it belongs in the projects that use it. I'd also rename `mkPerfCounters_Core` to just `mkPerfCounters`, with projects defining their own `mkPerfCounters_Synth` (or `mkPerfCounters_Foo` to avoid naming conflicts).

Reply:
So, I am keen for this to happen, but the reason we have it like this so far is that we claimed that "the library provides a synthesizable module". I am more than happy for that calim not to hold, and call it the consumer's problem to make it synthesizable. Anything that removes concrete values from the lib is  good. So far the halfway point we have is the allow the to flow through "-D". @JonasAlaif and I had looked at providing a macro to provide the synthsizable version of that module that consumer code could simply invoke with their concrete value and have access to a module constructor trivially. I do not remember why, but we ended up not doing it. @jrtc27 do you have thoughts on this approach?

Reply:
I've moved the instantiation now. We didn't create a macro since the definition of that was local to the file and could not be used in other files without a horrible `‘include` or `‘line`.